File: src/stemapp/web/js/peaks.js
// 波形 peaks（STPK 版 1、docs/PEAKS.md）の読み取りと、表示に使う解像度の選び方。

export const PEAKS_MAGIC = "STPK";
export const PEAKS_HEADER = 20;

/** ArrayBuffer を読む。形式が違えば Error。 */
export function parsePeaks(buf) {
  if (!(buf instanceof ArrayBuffer) || buf.byteLength < PEAKS_HEADER) {
    throw new Error("波形データが短すぎます。");
  }
  const view = new DataView(buf);
  const magic = String.fromCharCode(...new Uint8Array(buf, 0, 4));
  if (magic !== PEAKS_MAGIC) throw new Error("波形データの形式が違います。");
  const version = view.getUint16(4, true);
  if (version !== 1) throw new Error(`対応していない波形データの版です（${version}）。`);
  const samplesPerPx = view.getUint32(8, true);
  const sampleRate = view.getUint32(12, true);
  const points = view.getUint32(16, true);
  if (samplesPerPx <= 0 || sampleRate <= 0) throw new Error("波形データのヘッダが正しくありません。");
  if (buf.byteLength < PEAKS_HEADER + points * 2) throw new Error("波形データが途中で切れています。");
  const data = new Int8Array(buf, PEAKS_HEADER, points * 2); // data[2i]=min, data[2i+1]=max
  return { version, samplesPerPx, sampleRate, points, data };
}

/**
 * 曲全体（概観）用: 横幅 widthPx に曲全体を描くとき、1ピクセルに1点以上になる
 * いちばん細かい段階（点の数が widthPx 以下のうち samples_per_px が最小のもの）。
 * どれも多すぎるときはいちばん粗い段階。
 */
export function chooseOverviewLevel(levels, totalSamples, widthPx) {
  const sorted = [...levels].sort((a, b) => a - b);
  if (sorted.length === 0) return null;
  for (const spp of sorted) {
    if (Math.ceil(totalSamples / spp) <= Math.max(1, widthPx)) return spp;
  }
  return sorted[sorted.length - 1];
}

/**
 * 拡大表示用: 1ピクセルあたり wantedSpp サンプルで描くとき、
 * wantedSpp 以下でいちばん粗い段階（1ピクセルに1点以上）。細かすぎる表示ならいちばん細かい段階。
 */
export function chooseZoomLevel(levels, wantedSpp) {
  const sorted = [...levels].sort((a, b) => a - b);
  if (sorted.length === 0) return null;
  let best = sorted[0];
  for (const spp of sorted) {
    if (spp <= wantedSpp) best = spp;
  }
  return best;
}

/** 点 [i0, i1) の最小・最大（-1〜1）。範囲外は 0。 */
export function rangeMinMax(peaks, i0, i1) {
  const n = peaks.points;
  let lo = Math.max(0, Math.floor(i0));
  let hi = Math.min(n, Math.ceil(i1));
  if (hi <= lo) {
    if (lo >= n || i1 <= 0) return [0, 0];
    hi = lo + 1;
  }
  let mn = 127;
  let mx = -127;
  const d = peaks.data;
  for (let i = lo; i < hi; i++) {
    const a = d[2 * i];
    const b = d[2 * i + 1];
    if (a < mn) mn = a;
    if (b > mx) mx = b;
  }
  return [mn / 127, mx / 127];
}

File: src/stemapp/peaks.py
"""波形 peaks（波形表示用の事前計算データ）。形式は docs/PEAKS.md。

1ピクセル（samples_per_px サンプル）ごとに、左右を平均したモノラルの最小値・最大値を
int8（-127〜127）で持つ。最小値は切り下げ、最大値は切り上げで量子化するので、
量子化後の [min, max] は元の波形を必ず含む。
"""

from __future__ import annotations

import struct
from collections.abc import Iterable
from dataclasses import dataclass
from pathlib import Path

import numpy as np

MAGIC = b"STPK"
VERSION = 1
# マジック(4) 版(u16) 予約(u16) samples_per_px(u32) サンプルレート(u32) 点の数(u32)
HEADER = struct.Struct("<4sHHIII")
HEADER_SIZE = HEADER.size  # 20 バイト
SCALE = 127
DEFAULT_LEVELS: tuple[int, ...] = (256, 1024, 4096, 16384)


class PeaksError(ValueError):
    """peaks ファイルの形式が正しくない。"""


@dataclass(frozen=True)
class Peaks:
    samples_per_px: int
    sample_rate: int
    mins: np.ndarray  # int8
    maxs: np.ndarray  # int8

    @property
    def points(self) -> int:
        return int(self.mins.shape[0])


def _to_mono(data: np.ndarray) -> np.ndarray:
    arr = np.asarray(data, dtype=np.float32)
    if arr.ndim == 2:
        arr = arr.mean(axis=1, dtype=np.float32)
    return arr


def _quantize(mins: np.ndarray, maxs: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    qmin = np.clip(np.floor(mins.astype(np.float64) * SCALE), -SCALE, SCALE).astype(np.int8)
    qmax = np.clip(np.ceil(maxs.astype(np.float64) * SCALE), -SCALE, SCALE).astype(np.int8)
    return qmin, qmax


def compute_peaks(
    data: np.ndarray, sample_rate: int, levels: Iterable[int] = DEFAULT_LEVELS
) -> dict[int, Peaks]:
    """音声（(samples,) か (samples, 2)）から各解像度の peaks を作る。

    levels は小さい順に並べ直す。大きい解像度が小さい解像度の整数倍なら、小さい解像度の
    結果をまとめて計算する（結果は直接計算したものと同じ）。
    """
    mono = _to_mono(data)
    n = mono.shape[0]
    out: dict[int, Peaks] = {}
    prev_spp: int | None = None
    prev_min = prev_max = np.zeros(0, dtype=np.float32)
    for spp in sorted(set(int(v) for v in levels)):
        if spp <= 0:
            raise ValueError(f"samples_per_px は正の整数にしてください: {spp}")
        if n == 0:
            mins = maxs = np.zeros(0, dtype=np.float32)
        elif prev_spp is not None and spp % prev_spp == 0:
            step = spp // prev_spp
            idx = np.arange(0, prev_min.shape[0], step)
            mins = np.minimum.reduceat(prev_min, idx)
            maxs = np.maximum.reduceat(prev_max, idx)
        else:
            idx = np.arange(0, n, spp)
            mins = np.minimum.reduceat(mono, idx)
            maxs = np.maximum.reduceat(mono, idx)
        prev_spp, prev_min, prev_max = spp, mins, maxs
        qmin, qmax = _quantize(mins, maxs)
        out[spp] = Peaks(spp, sample_rate, qmin, qmax)
    return out


def encode_peaks(peaks: Peaks) -> bytes:
    body = np.empty(peaks.points * 2, dtype=np.int8)
    body[0::2] = peaks.mins
    body[1::2] = peaks.maxs
    header = HEADER.pack(MAGIC, VERSION, 0, peaks.samples_per_px, peaks.sample_rate, peaks.points)
    return header + body.tobytes()


def decode_peaks(blob: bytes) -> Peaks:
    if len(blob) < HEADER_SIZE:
        raise PeaksError("peaks ファイルが短すぎます。")
    magic, version, _reserved, spp, sr, points = HEADER.unpack_from(blob)
    if magic != MAGIC:
        raise PeaksError("peaks ファイルではありません（マジックが違います）。")
    if version != VERSION:
        raise PeaksError(f"未対応の peaks の版です: {version}")
    if len(blob) != HEADER_SIZE + points * 2:
        raise PeaksError("peaks ファイルの長さがヘッダの点の数と合いません。")
    body = np.frombuffer(blob, dtype=np.int8, offset=HEADER_SIZE)
    return Peaks(spp, sr, body[0::2].copy(), body[1::2].copy())


def write_peaks(path: Path, peaks: Peaks) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(encode_peaks(peaks))


def read_peaks(path: Path) -> Peaks:
    return decode_peaks(path.read_bytes())
